modules/backup-s3/client.go

```diff
@@ -61,37 +66,86 @@ func newClient(config *clientConfig, logger logrus.FieldLogger, dataPath string)
 	if err != nil {
 		return nil, errors.Wrap(err, "create client")
 	}
-	return &s3Client{client, config, logger, dataPath}, nil
+	return &s3Client{client, config, logger, dataPath, bucket, path, creds, region}, nil
+}
+
+func tempClient(currentClient *s3Client, overrideBucket, overridePath string, ctx context.Context) (*s3Client, error) {
+	AwsAccessKey := modulecomponents.GetValueFromContext(ctx, "X-AWS-ACCESS-KEY")
+	xAwsSecretKey := modulecomponents.GetValueFromContext(ctx, "X-AWS-SECRET-KEY")
+	xAwsSessionToken := modulecomponents.GetValueFromContext(ctx, "X-AWS-SESSION-TOKEN")
+	if overrideBucket != "" {
+		config := currentClient.config
+		creds := currentClient.creds
+		if AwsAccessKey != "" && xAwsSecretKey != "" {
+			creds = credentials.NewStaticV4(AwsAccessKey, xAwsSecretKey, xAwsSessionToken)
+		}
+
+		client, err := minio.New(config.Endpoint, &minio.Options{
+			Creds:  creds,
+			Region: currentClient.region,
+			Secure: config.UseSSL,
+		})
+		if err != nil {
+			return nil, errors.Wrap(err, "create minio client")
+		}
+		return &s3Client{client, config, currentClient.logger, currentClient.dataPath, overrideBucket, overridePath, creds, currentClient.region}, nil
+	} else if AwsAccessKey != "" || xAwsSecretKey != "" || xAwsSessionToken != "" {
+		return &s3Client{currentClient.client, currentClient.config, currentClient.logger, currentClient.dataPath, currentClient.bucket, currentClient.path, credentials.NewStaticV4(AwsAccessKey, xAwsSecretKey, xAwsSessionToken), currentClient.region}, nil
+	}
+	return currentClient, nil
```

Comment:
Can we refactor this method to a much simpler and readable version like this, bc with this code you are instantiating `s3Client` struct but all we need is to respect the header values (if they are present) and then create a new `minio client` and not return the whole `s3Client` struct again. Also with this approach we don't need `bucket` `path` and `creds` to be present in the `s3Client` struct, bc `bucket` and `path` are always dynamically defined and `creds` are not needed:
```
func (s *s3Client) getClient(ctx context.Context) (*minio.Client, error) {
	xAwsAccessKey := modulecomponents.GetValueFromContext(ctx, "X-AWS-ACCESS-KEY")
	xAwsSecretKey := modulecomponents.GetValueFromContext(ctx, "X-AWS-SECRET-KEY")
	xAwsSessionToken := modulecomponents.GetValueFromContext(ctx, "X-AWS-SESSION-TOKEN")
	if xAwsAccessKey != "" && xAwsSecretKey != "" && xAwsSessionToken != "" {
		return minio.New(s.config.Endpoint, &minio.Options{
			Creds:  credentials.NewStaticV4(xAwsAccessKey, xAwsSecretKey, xAwsSessionToken),
			Region: s.region,
			Secure: s.config.UseSSL,
		})
	}
	return s.client, nil
}
```

Reply:
I don't understand, can you show me how this works in the rest of the module?

Reply:
Such as the initialise method

Reply:
done